### PR TITLE
Historical Data Functionality EBBO Test 

### PR DESCRIPTION
**Pull request by shubhagarwal03**
First PR for EBBO Tool:

cow_endpoint_surplus.py can be run from test.py for historical data testing by specifying start and end blocks or even checking EBBO for a single hash, or from daemon_analytics.py for live monitoring where the program wakes every 30 mins and runs the current EBBO test on transactions in the period it was asleep. 
configuration.py holds a logging function, and certain abstracted values used in cow_endpoint_surplus. 
directory.py is used to fetch an etherscan api key from a config.py file in the root directory.

### CODE REVIEW

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [inline src/off-chain/historical-data/cowEndpointSurplus.py:35]: This should not be hardcoded in the class but rather part of the configuration file we read from.

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [inline src/off-chain/historical-data/cowEndpointSurplus.py:40]: Should be removed. This function should only calculate the surplus and not print the results.

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [inline src/off-chain/historical-data/cowEndpointSurplus.py:57]: I would organise the logic a bit differently:

1. Have an empty list of settlement hashes.
2. if settlement_hash is not None -> add a single hash to the list above.
3. elif start_block is not None and end_block is not None -> fetch all hashes and add them to the list above.
4. Once we have all the hashes, have a separate function that receives a list of hashes and returns a list of solver_competition data by calling cow api.
5. For each solver_competition data, call `get_order_surplus`

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [inline src/off-chain/historical-data/cowEndpointSurplus.py:106]: This function should not do any printing, but should return result that we print somewhere outside the function. Here, instead of printing, we can optionally log results.

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [inline src/off-chain/historical-data/cowEndpointSurplus.py:113]: `get_order_surplus` is too long and doing multiple things. We should split it. 

For example, there is no way to calculate the surplus based on the order_id. For this, we should create a function that does only one thing: calculates the surplus based on the minimum necessary order data. It should be a standalone function (not part of any class) and easily unit tested.

Then, we call this function from other functions...

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [inline src/off-chain/historical-data/cowEndpointSurplus.py:116]: We should not have pass the whole `soln` object if we only use `clearingPrices` out of it. Instead we should pass `clearingPrices`.

Also, this function is part of the `self` while also not using `self` anywhere. This is usually a sign that we need to move out the function from `EBBOHistoricalDataTesting`.

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [inline src/off-chain/historical-data/cowEndpointSurplus.py:175]: Do we have a unit test for `get_surplus_difference`

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [inline src/off-chain/historical-data/cowEndpointSurplus.py:116]: `buy_amount_from_id` rename to `buy_amount`

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [inline src/off-chain/historical-data/cowEndpointSurplus.py:117]: Rename to `sell_amount`

File: src/off-chain/historical-data/test.py
Comment (sunce86) [inline src/off-chain/historical-data/test.py:5]: Can we move this to `tests/unit/` folder so it get executed by CI?

**Review by sunce86 — COMMENTED** [review]
Nice. We are getting closer and closer.

Some comments mostly around code organization and making it more generic, reusable and complaint with unit tests.

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (shubhagarwal03) [reply src/off-chain/historical-data/cowEndpointSurplus.py:16]: If we're going to be logging/alerting eventually on a hash input basis, we could also get rid of the writing to a txt file functionality and just printing instead (as it was earlier).

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (harisang) [inline src/off-chain/historical-data/cowEndpointSurplus.py:62]: Not sure about it, but do we need to handle exceptions here with a try/except?

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (harisang) [inline src/off-chain/historical-data/cowEndpointSurplus.py:78]: Personal preference, just to avoid pushing the code more and more to the right due to indentation:

```
if order is liquidity:
    continue

```

Comment (harisang) [reply src/off-chain/historical-data/cowEndpointSurplus.py:78]: Also note that you have already counted such orders in the` total_orders` accounting.

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (shubhagarwal03) [reply src/off-chain/historical-data/cowEndpointSurplus.py:35]: Yep, took care of that in the code to be pushed.

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (shubhagarwal03) [reply src/off-chain/historical-data/cowEndpointSurplus.py:62]: Not required, but if we don't, we are assuming etherscan api will always return a valid result and will never be down. 
Not sure how likely/unlikely this is, but can be removed if needed.

Comment (shubhagarwal03) [reply src/off-chain/historical-data/cowEndpointSurplus.py:62]: This has been re-structured to fit Dusan's comment.
However, not required as such since we check whether a status code of the json returned is 200, if so, only then we add it to the list of competition data available for checking. Otherwise, it skips it and goes to the next hash.

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (shubhagarwal03) [reply src/off-chain/historical-data/cowEndpointSurplus.py:113]: _For this, we should create a function that does only one thing: calculates the surplus based on the minimum necessary order data._
Isn't this already get_surplus_difference, which uses minimum necessary data?

File: requirements.txt
Comment (bh2smith) [inline requirements.txt:5]: Looks like you are missing:

```
types-requests==2.28.11.15
```

File: src/off-chain/historical-data/cowEndpointSurplus.py
Comment (sunce86) [reply src/off-chain/historical-data/cowEndpointSurplus.py]: I will resolve this since the code is moved to a separate configuration file.

**Review by sunce86 — COMMENTED** [review]
Nice rework. I find the code easier to read now. Just a few more comments.

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (shubhagarwal03) [reply src/off_chain/historical_data/cow_endpoint_surplus.py:18]: yes, correct, this was changed after the logger. will delete this.

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (shubhagarwal03) [reply src/off_chain/historical_data/cow_endpoint_surplus.py:13]: Since this class now works for both historical and live monitoring, should I rename it to something like EBBOAnalysis?

**Comment by harisang** [comment]
Actually, could you at the very top add a description of what this PR is about, just for reference purposes?

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (harisang) [reply src/off_chain/historical_data/cow_endpoint_surplus.py:13]: Yeah, I would actually remove the historical part there. This is really the core of EBBO; it gets a block/block interval/hash, and it will report back the result. So, I agree we should get rename

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (shubhagarwal03) [reply src/off_chain/historical_data/cow_endpoint_surplus.py:91]: I had this in mind initially. However after testing over vast ranges, it seems that etherscan api picks up failed transactions as well for the cow settlement address. For example, 0x2497f7b9d7ebb1ee4ee069f26dc396257ca8030ac09d7af8626868dcea193b09
was picked up, but competition data not retrievable since transaction failed. Therefore, by checking status code of 200, we can filter out any failed transactions. Pls let me know if we can restructure how this is done.

Comment (harisang) [reply src/off_chain/historical_data/cow_endpoint_surplus.py:91]: Very good point. Haven't followed the prior discussion, but would it make sense to actually make the function only operate on a single hash, and then call it multiple times, once per hash, from within the `get_surplus_by_input` function?

Comment (harisang) [reply src/off_chain/historical_data/cow_endpoint_surplus.py:91]: > Therefore, by checking status code of 200, we can filter out any failed transactions.

And is it possible to distinguish between the entry not being existent in the database vs the api endpoint failing for whatever  reasons?

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (harisang) [inline src/off_chain/historical_data/cow_endpoint_surplus.py:95]: small typo in the comment "same rder"

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (harisang) [inline src/off_chain/historical_data/cow_endpoint_surplus.py:116]: If the API fails here, maybe we can add a logging message that says something like  "failed to check order uid=..."

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (shubhagarwal03) [reply src/off_chain/historical_data/cow_endpoint_surplus.py:91]: We were sort of doing this before, except for each hash we would compute the surplus directly in a big function. 
What you have suggested is certainly an alternative, but do we need to keep track of failed transactions?

File: src/off_chain/historical_data/daemon_analytics.py
Comment (harisang) [inline src/off_chain/historical_data/daemon_analytics.py:8]: A comment with a brief description of the main function is needed here

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (shubhagarwal03) [reply src/off_chain/historical_data/cow_endpoint_surplus.py:91]: to the second comment, yes there is actually. If the entry is not existent, status code = 404. If the comp. data endpoint fails for other reasons, it could be 500 (as seen a few weeks ago)

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (bh2smith) [reply src/off_chain/historical_data/cow_endpoint_surplus.py:95]:
```suggestion
    solutions that executed the same order and calculates surplus difference between that pair 
```

File: src/off_chain/historical_data/cow_endpoint_surplus.py
Comment (bh2smith) [inline src/off_chain/historical_data/cow_endpoint_surplus.py:105]: I find the fact that we are referring to solvers by name in a hard coded dictionary quite brittle. Do we really want to come back to this project and update the solver_dict manually every time a new solver is added? Would be better to use the `solver_address` field that is provided from the competition data and load the solver names from Dune via the API. Then everything is dynamic and only one place needs to be maintained (i.e. the solvers table in Dune). 

This is an [example query](https://dune.com/queries/1372857/2336383) showing solver details, but realistically it should be as simple as `select name, address from cow_protocol_ethereum.solvers`. Using the DuneAPI is also very straightforward.

File: src/off_chain/historical_data/daemon_analytics.py
Comment (bh2smith) [inline src/off_chain/historical_data/daemon_analytics.py:14]: What is the purpose of this sleep? are you trying to make `start_block != end_block`? 

I see another sleep at the top of the while loop that probably belongs in line 35 (immediately after the "going to sleep" log), but I am also not so sure this approach is ideal... there is a new EVM block published every 12 -15 seconds... this means that the program will wake up too often and start block will often equal end block. causing this entire block of code to run (and make requests) at least 12 times more than necessary. 

If sleep statement the way this is going to be approached I would suggest 60 seconds (~5 blocks).

**Review by bh2smith — COMMENTED** [review]
Took a quick look through and made some comments.

File: src/off_chain/historical_data/daemon_analytics.py
Comment (shubhagarwal03) [reply src/off_chain/historical_data/daemon_analytics.py:14]: I noticed that I pushed the sleep for one second (which is a debug test), but the final version is going to be asleep for 30 mins (1800 secs). I've added a comment at the top of main in the push (which I'm going to be making soon) that describes the exact behavior, but let me just put it here as well.

"""
_At runtime, this function records the most recent block and initializes it as the start block.
After the daemon is asleep for 30 mins, it gets the newest block as the end block. 
Since the competition endpoint has a lag of 30 mins (worst case), we wait 30 mins before we fetch comp. data
and start checking for potential surplus. 
If the data was not retrievable for any reason, it adds it to the list of unchecked hashes 
to be checked in the next cycle. Once all hashes in the current cycle have been iterated through, 
the previous end block + 1 becomes the start block for the next cycle, and the latest block is the new end block.
Daemon sleeps for 30 mins and continues checking._
"""

Comment (bh2smith) [reply src/off_chain/historical_data/daemon_analytics.py:14]: If its only every 30 minutes then it might make more sense for this to be a cronjob. That is, cheaper than running an idling service 24/7

Comment (harisang) [reply src/off_chain/historical_data/daemon_analytics.py:14]: I actually thought about the sleep time, in relation to some other discussion we also had with @bh2smith and it seems to me that we should actually reduce it after all to something like 5 minutes. This will ensure that we will most likely not get rate-limited from the competition endpoint, as the queries will be a few every few minutes instead of coming into bursts. It will also allow us to have better response times and get warnings faster, if needed.

Comment (bh2smith) [reply src/off_chain/historical_data/daemon_analytics.py:14]: Another suggestion here is to make the sleep time into a runtime configuration parameter.

File: src/daemon.py
Comment (harisang) [inline src/daemon.py:50]: Can this line just replace the right hand side of line 44?

File: src/daemon.py
Comment (harisang) [inline src/daemon.py:47]: I suspect we could remove lines 45 and 46 and write

`response_data = self.Instance.get_solver_competition_data([single_hash])`

Comment (shubhagarwal03) [reply src/daemon.py:48]: yes, thank you.

File: src/off_chain/cow_endpoint_surplus.py
Comment (shubhagarwal03) [reply src/off_chain/cow_endpoint_surplus.py:293]: I've changed it to something like this:
        for key, value in self.solver_dict.items():
            if value[0] == 0:

File: src/daemon.py
Comment (shubhagarwal03) [reply src/daemon.py:35]: Each sleep time was set to 30 mins so that once start_block and end_block have a value, cow endpoint is available in the first iteration since we waited 30 mins. But again, it can be altered to this:

startblock = w3.eth.block_number
while True:
   time.sleep(30 mins)
   endblock = w3.eth.block_number
   .... do stuff
   startblock = endblock + 1
   
   In this case however, for the first few iterations, all the hashes will go to unchecked_hashes since data from endpoint is not retrievable.

**Review by bh2smith — COMMENTED** [review]
Just a few things I noticed when passing through.

File: .pylintrc
Comment (shubhagarwal03) [reply .pylintrc]: Had borrowed some from your solver-template pylint. 
These are the ignores now:
fixme,too-few-public-methods,too-many-arguments,too-many-locals,pointless-string-statement,import-error

Comment (bh2smith) [reply .pylintrc]: Among the remaining, I would think that 

- `too-many-locals` indicates you have code blocks that have become too large and are likely not focused on accomplishing a single task, thus many variables are being created within the function. I would try to resolve this.
- `pointless-string-statement` also looks like code that could just be removed.
- `import-error` should probably also not be silenced.

**Comment by bh2smith** [comment]
Note -- once fixes are made, you can "resolve" the comments and it will be a lot easier to look at everything here.

File: .pylintrc
Comment (shubhagarwal03) [reply .pylintrc]: how do you ignore the secret key file, import-error was ignored since keys file isn't pushed to github:
https://github.com/cowprotocol/ebbo/actions/runs/4683373799/jobs/8298384068?pr=10

File: .github/workflows/deploy.yaml
Comment (harisang) [inline .github/workflows/deploy.yaml:39]: Would be good to move this to a separate PR. It's a self-contained functionality, and once we merge this two-line PR, then everything else should be easier.

File: .pylintrc
Comment (bh2smith) [reply .pylintrc]: I would think that if the project is relying on a file that is not comitted to the project, then it sounds like you should be putting those "secrets" into a dotenv file (i.e. a file `.env`) that contains a list of constants like 

```
DUNE_API_KEY=
INFRUA_KEY=
```

Then you can `load_dotenv()` and read the environment variables you need in this way rather than an "import" statement.

Here is some docs (its very easy to use) https://pypi.org/project/python-dotenv/

```
pip install python-dotenv
```

and maybe something like this is the flavour you are looking for:

```py
from dotenv import dotenv_values

config = dotenv_values(".env")  # config = {"USER": "foo", "EMAIL": "foo@example.org"}
```

File: tests/unit/test.py
Comment (bh2smith) [inline tests/unit/test.py]: I know its hard to let go, but... maybe best to remove dead code.

**Review by harisang — APPROVED** [review]
Approving so that we get things going. And a clean-up or minor iterations can be done in a follow-up PR.

**Pull request merged**
